### DISCUSSION TITLE
How to query/update a self-relation?

### DISCUSSION THREAD

**Discussion by dXu23** [prisma/prisma#12310] [Q&A] [answered]
I have some self-relation tables taken directly from one of Prisma's [examples](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/self-relations#many-to-many-self-relations)

```
model User {
  id         Int       @id @default(autoincrement())
  name       String?
  followedBy Follows[] @relation("follower")
  following  Follows[] @relation("following")
}

model Follows {
  follower    User @relation("follower", fields: [followerId], references: [id])
  followerId  Int
  following   User @relation("following", fields: [followingId], references: [id])
  followingId Int

  @@id([followerId, followingId])
}
```
I am wondering on the best/most idiomatic way to determine whether a user is following another user, as well as operations for following and unfollowing. I'm thinking the best way is to just query/update the Follows table directly, but I'm unsure whether the followedBy and following fields in the User model will automatically update themselves.

**Comment by nurul3101** [comment, marked as answer]
Hey @dXu23 👋 

Consider this example:

Creating a user:
```javascript
  const createUsers = await prisma.user.createMany({
    data: [
      {
        name: 'user1',
      },
      {
        name: 'user2',
      },
      {
        name: 'user3',
      },
    ],
  });
```
Defining follows relationship:
```javascript
  const followUsers = await prisma.follows.createMany({
    data: [
      {
        followerId: 1,
        followingId: 2,
      },
      {
        followerId: 1,
        followingId: 3,
      },
      {
        followerId: 2,
        followingId: 3,
      },
    ],
  });
```
To fetch who are the followers of user 1, it can be achieved through user model as well as follows model as described below
```javascript
  const findFollowersThroughUser = await prisma.user.findMany({
    where: {
      following: {
        some: {
          followerId: 1,
        },
      },
    },
  });

  console.log('findFollowersThroughUser', findFollowersThroughUser);

  const findFollowersThroughFollow = await prisma.follows.findMany({
    where: {
      followerId: 1,
    },
    include: {
      following: true,
    },
  });

  console.log('findFollowersThroughFollow', findFollowersThroughFollow);
```  
Here's the output:
```javascript
findFollowersThroughUser [ { id: 2, name: 'user2' }, { id: 3, name: 'user3' } ]

findFollowersThroughFollow [
  {
    followerId: 1,
    followingId: 2,
    following: { id: 2, name: 'user2' }
  },
  {
    followerId: 1,
    followingId: 3,
    following: { id: 3, name: 'user3' }
  }
]
```
For updating the relations when someone unfollows, the record needs to be explicitly removed from the `follows` table in this case.

**Reply by giuseppeg** [reply]
Hi @nurul3101, given @dXu23's model and a `User.id` how would you find (and count) the users that are following each other with `User.id`?

**Reply by nurul3101** [reply]
Solution Discussed here: https://stackoverflow.com/questions/74436507/find-users-that-follow-each-other-on-a-prisma-self-relation/74443057

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
